### PR TITLE
Convert android to kotlin

### PR DESCRIPTION
**Pull request by johnbotris**

### CODE REVIEW

File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmModel.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		interval: Int,
+		endType: EndType?,
+		endValue: Long,
+		alarmTrigger: AlarmTrigger,
+		localTimeZone: TimeZone,
+		callback: AlarmIterationCallback,
+	) {
+		val isAllDayEvent = isAllDayEventByTimes(eventStart, eventEnd)
+		val calcEventStart = if (isAllDayEvent) getAllDayDateLocal(eventStart, localTimeZone) else eventStart
+		val endDate = if (endType == EndType.UNTIL) if (isAllDayEvent) getAllDayDateLocal(Date(endValue), localTimeZone) else Date(endValue) else null
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmModel.kt:25]: nested if in one line :exploding_head:

File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmModel.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		utcCalendar[calendar[Calendar.YEAR], calendar[Calendar.MONTH], calendar[Calendar.DAY_OF_MONTH], 0, 0] = 0
+		utcCalendar[Calendar.MILLISECOND] = 0
+		return utcCalendar.time
+	}
+
+	private fun getAllDayDateLocal(utcDate: Date, localTimeZone: TimeZone): Date {
+		val utcCalendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
+		utcCalendar.time = utcDate
+		val calendar = Calendar.getInstance(localTimeZone)
+		calendar[utcCalendar[Calendar.YEAR], utcCalendar[Calendar.MONTH], utcCalendar[Calendar.DAY_OF_MONTH], 0, 0] = 0
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmModel.kt:98]: this is a very cursed set call

**Comment by charlag** [comment]
Also tests are borked

File: app-android/app/src/main/java/de/tutao/tutanota/push/TutanotaNotificationsHandler.kt [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+			val responseCode = urlConnection.responseCode
+			Log.d(TAG, "MissedNotification response code $responseCode")
+			handleResponseCode(urlConnection, responseCode)
+			urlConnection.inputStream.use { inputStream ->
+				val responseString = IOUtils.toString(inputStream, StandardCharsets.UTF_8)
+				Log.d(TAG, "Loaded Missed notifications response")
+				return MissedNotification.fromJson(JSONObject(responseString))
+			}
+		} catch (e: MalformedURLException) {
+			throw RuntimeException(e)
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/push/TutanotaNotificationsHandler.kt:136]: no need for rewrap

File: app-android/app/src/main/java/de/tutao/tutanota/push/TutanotaNotificationsHandler.kt [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+	private fun extractSuspensionTime(urlConnection: HttpURLConnection): Int {
+		var retryAfterHeader = urlConnection.getHeaderField("Retry-After")
+		if (retryAfterHeader == null) {
+			retryAfterHeader = urlConnection.getHeaderField("Suspension-Time")
+		}
+		return try {
+			retryAfterHeader!!.toInt()
+		} catch (e: NumberFormatException) {
+			0
+		}
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/push/TutanotaNotificationsHandler.kt:174]:
```suggestion
		return retryAfterHeader.toIntOrNull() ?: 0
```

File: app-android/build.gradle [hunk 5]
```diff
@@ -1,13 +1,15 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
+	ext.kotlin_version = '1.4.10'
```
Comment (charlag) [inline app-android/build.gradle:4]: Lol? How about 1.6.21?

**Comment by charlag** [comment]
Need to port bugfix for `CredentialsEncryptionFromAPI30`

File: app-android/app/src/main/java/de/tutao/tutanota/Utils.kt [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+	fun fileToUri(file: File?): String {
+		return Uri.fromFile(file).toString()
+	}
+
+	@SuppressLint("Range")
+	@Throws(FileNotFoundException::class)
+	fun getFileInfo(context: Context, fileUri: Uri): FileInfo {
+		val scheme = fileUri.scheme
+		if (scheme == null || scheme == "file") {
```
[the remark this answers is not in the document]
Comment (johnbotris) [reply app-android/app/src/main/java/de/tutao/tutanota/Utils.kt:59]: there are two ways to end up at the `throw RuntimeException(...)`, I might just leave for now

**Comment by johnbotris** [comment]
native contact search in the mail editor throws an uncaught SecurityException when contact permission isn't granted. We only handle `PermissionError` in the web side. Not sure what got lost where. will fix it once we merge bubble text fields rewrite.

File: app-android/app/src/main/java/de/tutao/tutanota/Native.kt [hunk 7]
```diff
@@ -347,21 +359,23 @@ class Native internal constructor(
...
+				TAG, """
      Error attaching thumbnail to share intent:
      ${e.message}
-     """.trimIndent())
+     """.trimIndent()
+			)
 		}
 		sendIntent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
 		val intent = Intent.createChooser(sendIntent, null)
 		activity.startActivity(intent)
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/Native.kt:377]: you are running the whole function on `IO` dispatcher but this part should def be on `Main`

File: app-android/app/src/main/java/de/tutao/tutanota/Utils.kt [hunk 8]
```diff
@@ -7,158 +7,141 @@ import android.os.Build
...
+	// alpha channel is always max
+	return rgb or -0x1000000
+}
 
-		// Counting the perceptive luminance
-		// human eye favors green color...
-		val a = 1 - (0.299 * r + 0.587 * g + 0.114 * b) / 255
-		return a < 0.5
-	}
+class IterableBufferedReader(reader: Reader) : BufferedReader(reader), Iterable<String> {
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/Utils.kt:129]: maybe add a test

File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmBroadcastReceiver.kt [hunk 9]
```diff
@@ -6,24 +6,30 @@ import android.content.Intent
...
 	}
 
 	companion object {
 		const val ALARM_NOTIFICATION_CHANNEL_ID = "alarms"
 		private const val TAG = "AlarmBroadcastReceiver"
 		private const val SUMMARY_EXTRA = "summary"
 		const val EVENT_DATE_EXTRA = "eventDate"
-		fun makeAlarmIntent(context: Context?, occurrence: Int, identifier: String, summary: String?, eventDate: Date?, userId: String?): Intent {
+		fun makeAlarmIntent(
+			context: Context?,
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmBroadcastReceiver.kt:26]: should not be nullable

File: app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmBroadcastReceiver.kt [hunk 10]
```diff
@@ -6,24 +6,30 @@ import android.content.Intent
...
+			occurrence: Int,
+			identifier: String,
+			summary: String?,
+			eventDate: Date?,
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/alarms/AlarmBroadcastReceiver.kt:30]: we unwrap `EventDate`

File: app-android/app/src/main/java/de/tutao/tutanota/alarms/SystemAlarmFacade.kt [hunk 11]
```diff
@@ -30,9 +32,12 @@ class SystemAlarmFacade(private val context: Context) {
 	private val alarmManager: AlarmManager
 		get() = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
 
-	private fun makeAlarmPendingIntent(occurrence: Int, identifier: String, summary: String?,
-									   eventDate: Date?, user: String?): PendingIntent {
-		val intent: Intent = AlarmBroadcastReceiver.makeAlarmIntent(context, occurrence, identifier, summary, eventDate, user)
+	private fun makeAlarmPendingIntent(
+		occurrence: Int, identifier: String, summary: String?,
```
Comment (charlag) [inline app-android/app/src/main/java/de/tutao/tutanota/alarms/SystemAlarmFacade.kt:37]: hm probably some params not nullable?

File: src/api/common/utils/Utils.ts [hunk 12]
```diff
@@ -149,8 +149,12 @@ export function isSecurityError(e: any): boolean {
 }
 
 export function objToError(o: Record<string, any>): Error {
+	console.log(o)
```
Comment (charlag) [inline src/api/common/utils/Utils.ts:152]: ??

**Review by charlag — APPROVED** [review]

**Pull request merged**